### PR TITLE
imageService: Rename ExportImage, LoadImage, SaveImage

### PR DESCRIPTION
**Pull request by vvoland**
**- What I did**

- Renamed `ExportImage` to `SaveImage` so it matches the cli operation. Previously `docker save` ended up in daemon's `ExportImage` which is easily confused with `docker export`.
Note: `docker export` doesn't have a corresponding function in the `ImageService` interface yet. I'll prepare a follow-up for that, it will also allow to share common bits with future containerd implementation.

- Renamed `LoadImage` to `LoadImages` and `SaveImage` to `SaveImages` because these functions can import more than one image.

- Renamed `ImportImage` to `ImportImageRootFS` to be clear that we are importing an image from rootfs only.

**- How I did it**
with vim

**- How to verify it**
Check if build didn't break

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by rumpl** [comment]
`SaveImage exports a list of images to the given output stream.` < shouldn't it then be called `SaveImages` ?

**Comment by vvoland** [comment]
Even with multiple source images, we still produce only one output tar archive - one image.
But yeah, I think it makes more sense to look at it from perspective of `saving (from) images` and not `saving (to) image`.
For consistency `LoadImage` would also need to be changed to `LoadImages`.

File: api/server/router/image/backend.go [hunk 1]
```diff
@@ -31,9 +31,9 @@ type imageBackend interface {
 }
 
 type importExportBackend interface {
-	LoadImage(ctx context.Context, inTar io.ReadCloser, outStream io.Writer, quiet bool) error
-	ImportImage(ctx context.Context, ref reference.Named, platform *specs.Platform, msg string, layerReader io.Reader, changes []string) (dockerimage.ID, error)
-	ExportImage(ctx context.Context, names []string, outStream io.Writer) error
+	LoadImages(ctx context.Context, inTar io.ReadCloser, outStream io.Writer, quiet bool) error
```
Comment (thaJeztah) [inline api/server/router/image/backend.go:34]: We _really_ need to deprecate that `quiet` option 😞. It should _never_ have been part of the API. I noticed that one some time ago (haven't checked exactly how it's used anywhere), but we can even consider to

- remove the option from the backend (instead making the API discard the output if it's set?)
- deprecate (gated by API version), and make sure the _client_ discards the output

**Comment by thaJeztah** [comment]
Thanks! So, looking at this, perhaps we should have a look at the overall picture. I still think we should look at renaming some of these, but as we're changing interfaces, perhaps this is also the right moment to get a better insight into this _overall_, instead of making random improvements "here and there". We can still make changes granularly, but it would be good to have a "northern star" (or at least a consistent approach) for these. 

I only had a first (cursory) glance after checking out the code, but some thoughts;

### Naming: "no image prefix/suffix", "prefix", or "suffix"?

As the package is named `image` (so we're dealing with .. images), my first thought was; we should remove the `Image` part from the functions, so just `Load()`, `Import()`, `Save()`; effectively it's then `image.Load()`, `img.Save()` etc. :point_left: This would be my preference (and most idiomatic) _if possible_.

_However_ (this is what I didn't check); that _may_ be confusing if this interface is implemented in a backend that deals with more than just _images_, so if that's the case (i.e., if we must have something "unique" in these names), we should look for consistency with other backends. Unfortunately, it looks like (erm) there isn't much consistency'; a quick glance; 

Plugin backend: https://github.com/moby/moby/blob/219ced474d8bcb9cbd399a6ab6ec17f178fe3306/api/server/router/plugin/backend.go#L16-L21

Container backend; https://github.com/moby/moby/blob/219ced474d8bcb9cbd399a6ab6ec17f178fe3306/api/server/router/container/backend.go#L34-L43

Network backend; https://github.com/moby/moby/blob/219ced474d8bcb9cbd399a6ab6ec17f178fe3306/api/server/router/network/backend.go#L14-L18

Or worse, backends that use _both_ https://github.com/moby/moby/blob/219ced474d8bcb9cbd399a6ab6ec17f178fe3306/api/server/router/volume/backend.go#L15-L21 https://github.com/moby/moby/blob/219ced474d8bcb9cbd399a6ab6ec17f178fe3306/api/server/router/volume/backend.go#L27-L34


### Options

This is _definitely_ something we should look at; 

1. Some backends pass _all_ options as individual arguments. This (IMO) is not desirable;
    - Adding an option requires breaking the interface
    - No sharing between client and server (effectively, client and server could be considered "symmetrical"; client calls API endpoint with options -> serializes across the wire -> server decodes options)
    - Hard to discover what options are available, and documenting options must be done in multiple places (which easy gets out of sync)
2. Some backends use option _structs_
    - Can be shared between the client and server (e.g., see https://github.com/moby/moby/commit/e8fa708ae5c460ebeec1a88a3ec3615272efaa4f)
    - ^^ Which also can help with options that are API version dependent (both client and server can share same logic for "omitting" options that are not supported by the API version used)
    - Helps discoverability (these are the options I can use; fields can be documented)
    - Easy to construct (initialize a struct; set all options we need to set)
3. Some backends use functional arguments (e.g. [Volumes Backend](https://github.com/moby/moby/blob/219ced474d8bcb9cbd399a6ab6ec17f178fe3306/api/server/router/volume/backend.go#L17-L19))
    - Functional arguments can be powerful and flexible (see containerd)
    - Not always easy to discover
    - May result in lots of boiler-plating code (whole function to set a boolean)
    - If functional arguments depends on action (e.g., only for "create", not for "update"), we need a way to restrict their use; this may mean "duplicate all of the functions" (`CreateWithSomeBool`, `UpdateWithSomeBool`) 

My preference would go to either `2.` or `3.`. But we can even consider that `2.` and `3.` are not "mutually exclusive". For example;

```go
func WithStopOptions(opts StopOptions) error {}
func WithStopTimeOut(seconds int) error {}
func WithStopSignal(signal string) error {}
```

In the above, either individual options can be passed

```go
containers.Stop(ctx, containerID, WithStopTimeout(10), WithStopSignal("SIGTERM"))
```

Or all options at once with a struct;

```go
containers.Stop(ctx, containerID, WithStopOptions(container.StopOptions{
    Signal: "SIGTERM",
    Timeout: 10,
}))
```

### TL;DR

:point_up: perhaps the best course of action is to;

- create a list of all the backend interfaces we have
- (also have a quick look if they're used "in combination", which may be relevant w.r.t. requiring pre/suffixes)
- find common patterns (not all things will have counterparts in other object types, but most would have at least `CRUD` methods)
- pick our preferred pattern / conventions, and document that somewhere (can be a tracking ticket)

:warning: when looking at this, we should also consider that the "backend" does not _have to be_ a 1-1 mapping to "endpoints". For example, a "ImportRootFS" can also be considered a `construct image from rootfs --> import/load image`. We need to look what abstraction makes most sense, and _where_ that abstraction should happen.
